### PR TITLE
SONARJNKNS-217 Allow to reuse global server configuration in any buil…

### PR DESCRIPTION
**Pull request by dbmeneses**
…d step

### CODE REVIEW

File: src/test/java/hudson/plugins/sonar/SonarBuildWrapperTest.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  @Test
+  public void testBuild() throws Exception {
+    // set up a free style project with our wrapper that will execute CaptureVarsBuilder
+    SonarBuildWrapper wrapper = new SonarBuildWrapper("local");
+    configureSonar(createTestInstallation());
+    CaptureVarsBuilder b = new CaptureVarsBuilder();
+    FreeStyleProject project = setupFreeStyleProject(b);
+    project.getBuildWrappersList().add(wrapper);
+    Run<?, ?> build = build(project);
+
```
Comment (SonarTech) [inline src/test/java/hudson/plugins/sonar/SonarBuildWrapperTest.java]: CRITICAL Complete the assertion. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS2970)

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+  @Nullable
+  public SonarInstallation getInstallation() {
+    return SonarInstallation.get(getInstallationName());
+  }
+
+  @Override
+  public Environment setUp(AbstractBuild build, Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
+    if (!SonarRunnerBuilder.isSonarInstallationValid(getInstallationName(), listener)) {
+      return new Environment() {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: MAJOR Make this a named "static" inner class. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS2694)

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+      };
+    }
+    return new SonarEnvironment(getInstallation(), listener.getLogger());
+  }
+
+  @Extension
+  public static final class DescriptorImpl extends BuildWrapperDescriptor
+  {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: MAJOR Move this left curly brace to the end of previous line of code. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3ALeftCurlyBraceEndLineCheck)

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+      return SonarInstallation.all();
+    }
+
+    @Override
+    public String getHelpFile() {
+      return "/plugin/sonar/help-buildWrapper.html";
+    }
+  }
+
+  class SonarEnvironment extends Environment {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: MAJOR Make this a "static" inner class. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS2694)

File: src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+      int nbMaskedPasswords = 0;
+      for (String password : passwords) {
+        if (StringUtils.isNotEmpty(password)) { // we must not handle empty passwords
+          regex.append(Pattern.quote(password));
+          regex.append('|');
+          nbMaskedPasswords++;
+        }
+      }
+      if (nbMaskedPasswords >= 1) { // is there at least one password to mask?
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3ATrailingCommentCheck)

File: src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        regex.deleteCharAt(regex.length() - 1); // removes the last unuseful pipe
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3ATrailingCommentCheck)

File: src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    }
+  }
+
+  @Override
+  protected void eol(byte[] bytes, int len) throws IOException {
+    String line = new String(bytes, 0, len);
+    if (passwordsAsPattern != null) {
+      line = passwordsAsPattern.matcher(line).replaceAll(REPLACEMENT);
+    }
+    logger.write(line.getBytes());
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java]: MAJOR Remove this use of "getBytes" [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS1943)

File: src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Collection;
+import java.util.regex.Pattern;
+
+/**
+ * Class taken from the mask-passwords plugin
+ */
+public class MaskPasswordsOutputStream extends LineTransformationOutputStream {
+  private final static String REPLACEMENT = "******";
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java]: MINOR Reorder the modifiers to comply with the Java Language Specification. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AModifiersOrderCheck)

File: src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        passwordsAsPattern = null;
+      }
+    } else { // no passwords to hide
+      passwordsAsPattern = null;
+    }
+  }
+
+  @Override
+  protected void eol(byte[] bytes, int len) throws IOException {
+    String line = new String(bytes, 0, len);
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java]: MAJOR Remove this use of constructor "String(byte[],int,int)" [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS1943)

File: src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        }
+      }
+      if (nbMaskedPasswords >= 1) { // is there at least one password to mask?
+        regex.deleteCharAt(regex.length() - 1); // removes the last unuseful pipe
+        regex.append(')');
+        passwordsAsPattern = Pattern.compile(regex.toString());
+      } else { // no passwords to hide
+        passwordsAsPattern = null;
+      }
+    } else { // no passwords to hide
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3ATrailingCommentCheck)

File: src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+public class MaskPasswordsOutputStream extends LineTransformationOutputStream {
+  private final static String REPLACEMENT = "******";
+  private final OutputStream logger;
+  private final Pattern passwordsAsPattern;
+
+  public MaskPasswordsOutputStream(OutputStream logger, Collection<String> passwords) {
+
+    this.logger = logger;
+
+    if (passwords != null && passwords.size() > 0) {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java]: MAJOR Use isEmpty() to check whether the collection is empty or not. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS1155)

File: src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+      // passwords are aggregated into a regex which is compiled as a pattern
+      // for efficiency
+      StringBuilder regex = new StringBuilder().append('(');
+
+      int nbMaskedPasswords = 0;
+      for (String password : passwords) {
+        if (StringUtils.isNotEmpty(password)) { // we must not handle empty passwords
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3ATrailingCommentCheck)

File: src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+          regex.append(Pattern.quote(password));
+          regex.append('|');
+          nbMaskedPasswords++;
+        }
+      }
+      if (nbMaskedPasswords >= 1) { // is there at least one password to mask?
+        regex.deleteCharAt(regex.length() - 1); // removes the last unuseful pipe
+        regex.append(')');
+        passwordsAsPattern = Pattern.compile(regex.toString());
+      } else { // no passwords to hide
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/MaskPasswordsOutputStream.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3ATrailingCommentCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 14 issues:
* CRITICAL 1 critical
* MAJOR 7 major
* MINOR 6 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'sonar' package (org.jenkins-ci.plugins:sonar:target/generated-sources/localizer/hudson/plugins/sonar) [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 13 issues:
* MAJOR 7 major
* MINOR 6 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'sonar' package (org.jenkins-ci.plugins:sonar:target/generated-sources/localizer/hudson/plugins/sonar) [rule](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by dbmeneses** [comment]
I will have a look at all SQ issues separately, to avoid rebasing again all P/R

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+      map.put("SONAR_JDBC_USERNAME", getOrDefault(inst.getDatabaseLogin(), ""));
+      map.put("SONAR_JDBC_PASSWORD", getOrDefault(inst.getDatabasePassword(), ""));
+
+      if (StringUtils.isEmpty(inst.getMojoVersion())) {
+        map.put("SONAR_MAVEN_GOAL", "sonar:sonar");
+      } else {
+        map.put("SONAR_MAVEN_GOAL", "org.codehaus.mojo:sonar-maven-plugin:" + installation.getMojoVersion() + ":sonar");
+      }
+
+      map.put("SONAR_ADDITIONAL", getOrDefault(inst.getAdditionalProperties(), ""));
```
Comment (henryju) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: Maybe rename SONAR_EXTRA_PROPS?

**Comment by henryju** [comment]
LGTM

**Pull request merged**